### PR TITLE
GUACAMOLE-41: Add LDAP section to readme.

### PR DESCRIPTION
**Pull request by jmuehlner**

### CODE REVIEW

File: guacamole-docker/README.md [hunk 1]
```diff
@@ -115,6 +119,35 @@ The process for doing this via the `mysql` utility included with MySQL is
 documented in
 [the Guacamole manual](http://guacamole.incubator.apache.org/doc/gug/jdbc-auth.html#jdbc-auth-mysql).
 
+
+Deploying Aoache Guacamole with LDAP authentication
```
Comment (mike-jumper) [inline guacamole-docker/README.md]: Apache*

File: guacamole-docker/README.md [hunk 2]
```diff
@@ -115,6 +119,35 @@ The process for doing this via the `mysql` utility included with MySQL is
...
+        -e LDAP_HOSTNAME=172.17.42.1                            \
+        -e LDAP_USER_BASE_DN=ou=people,dc=example,dc=com        \
+        -e LDAP_CONFIG_BASE_DN=ou=connections,dc=example,dc=com \
+        -d -p 8080:8080 glyptodon/guacamole
+
+Linking Guacamole to your LDAP directory will require additional configuration parameters
+specified via environment variables. These variables collectively describe how Guacamole
+will connect to LDAP:
+
+1. `LDAP_HOSTNAME` - The base of the DN for all Guacamole users.
```
Comment (mike-jumper) [inline guacamole-docker/README.md]: Nope. Copy/paste error, I assume?

File: guacamole-docker/README.md [hunk 3]
```diff
@@ -115,6 +119,35 @@ The process for doing this via the `mysql` utility included with MySQL is
...
+Deploying Aoache Guacamole with LDAP authentication
+--------------------------------------------------
+
+    docker run --name some-guacamole --link some-guacd:guacd    \
+        -e LDAP_HOSTNAME=172.17.42.1                            \
+        -e LDAP_USER_BASE_DN=ou=people,dc=example,dc=com        \
+        -e LDAP_CONFIG_BASE_DN=ou=connections,dc=example,dc=com \
+        -d -p 8080:8080 glyptodon/guacamole
+
+Linking Guacamole to your LDAP directory will require additional configuration parameters
```
Comment (mike-jumper) [inline guacamole-docker/README.md]: I think technically it's not "linking" in this case, since Docker links are not involved (no `--link`).

File: guacamole-docker/README.md [hunk 4]
```diff
@@ -115,6 +119,35 @@ The process for doing this via the `mysql` utility included with MySQL is
...
+specified via environment variables. These variables collectively describe how Guacamole
+will connect to LDAP:
+
+1. `LDAP_HOSTNAME` - The base of the DN for all Guacamole users.
+2. `LDAP_USER_BASE_DN` - The base of the DN for all Guacamole users.
+3. `LDAP_PORT` - The port your LDAP server listens on.  (Optional)
+4. `LDAP_ENCRYPTION_METHOD` - The encryption mechanism that Guacamole should use when
+communicating with your LDAP server. (Optional)
+5. `LDAP_GROUP_BASE_DN` - (Optional) The base of the DN for all groups that may be
```
Comment (mike-jumper) [inline guacamole-docker/README.md]: Why does this one start with "(Optional)" while all the others have "(Optional)" at the end?

File: guacamole-docker/README.md [hunk 5]
```diff
@@ -12,8 +12,12 @@ How to use this image
 =====================
 
 Using this image will require an existing, running Docker container with the
-[guacd image](https://registry.hub.docker.com/u/glyptodon/guacd/), and another
-Docker container providing either a PostgreSQL or MySQL database.
+[guacd image](https://registry.hub.docker.com/u/glyptodon/guacd/), and either
+network access to a working LDAP server, or another Docker container providing
+a PostgreSQL or MySQL database.
+
+SQL Based Authentication
```
Comment (mike-jumper) [inline guacamole-docker/README.md]: Is this header necessary?

I have nothing against the header itself, but it seems odd that the MySQL, PostgreSQL, *and* the LDAP configurations all fall beneath it.

File: guacamole-docker/README.md (hunk 3 above)
Comment (mike-jumper) [inline guacamole-docker/README.md]: Looks like I made this mistake, too, within the manual:

~~~xml
<para>Linking Guacamole to your LDAP directory will require additional configuration
  parameters specified via environment variables. These variables collectively
  describe how Guacamole will connect to LDAP:</para>
~~~

Unless you think the wording is OK as-is.

**Comment by mike-jumper** [comment]
Are we going to describe the combination of LDAP with a database as well?

File: guacamole-docker/README.md (hunk 4 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: Oops! I'll fix that.

File: guacamole-docker/README.md (hunk 5 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: I added that because the paragraphs below are specific to the SQL extensions. I can remove it if it seems confusing.

File: guacamole-docker/README.md (hunk 1 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: Oops

File: guacamole-docker/README.md (hunk 2 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: Yeah, lemme fix that.

File: guacamole-docker/README.md (hunk 3 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: We should probably fix it in both places... "Using Guacamole with your LDAP directory"?

**Comment by jmuehlner** [comment]
Yeah, we should probably mention that both can be used together as well.

File: guacamole-docker/README.md (hunk 5 above)
Comment (mike-jumper) [inline guacamole-docker/README.md]: The main reason I find this odd is that the outline of the document, taking header level into account, is now the following:

1. What is Apache Guacamole?
2. How to use this image
3. SQL Based Authentication
    1. Deploying Guacamole with PostgreSQL authentication
    2. Deploying Guacamole with MySQL authentication
    3. Deploying Aoache Guacamole with LDAP authentication

Comment (jmuehlner) [inline guacamole-docker/README.md]: Yeah, it's probably best if I just remove that header, or make it a sub-header.

Comment (mike-jumper) [inline guacamole-docker/README.md]: If you do decided to keep it, I'd suggest rewording to "Database authentication". That would be consistent with the other existing docs (the manual), and consistent with the style of the other headers in this document (sentence case rather than title case).

File: guacamole-docker/README.md [hunk 6]
```diff
@@ -120,7 +117,7 @@ documented in
 [the Guacamole manual](http://guacamole.incubator.apache.org/doc/gug/jdbc-auth.html#jdbc-auth-mysql).
 
 
-Deploying Aoache Guacamole with LDAP authentication
+Deploying Apache Guacamole with LDAP and database authentication
```
Comment (mike-jumper) [inline guacamole-docker/README.md]: But this is just with LDAP

File: guacamole-docker/README.md [hunk 7]
```diff
@@ -148,6 +145,24 @@ binding as `LDAP_SEARCH_BIND_DN` to authenticate other users. (Optional)
 username within all Guacamole user objects in the LDAP directory. (Optional)
 9. `LDAP_CONFIG_BASE_DN` - The base of the DN for all Guacamole configurations. (Optional)
 
+Deploying Apache Guacamole with LDAP authentication
```
Comment (mike-jumper) [inline guacamole-docker/README.md]: But this is LDAP + database

File: guacamole-docker/README.md [hunk 8]
```diff
@@ -129,17 +126,17 @@ Deploying Aoache Guacamole with LDAP authentication
...
 
-Linking Guacamole to your LDAP directory will require additional configuration parameters
+Using Guacamole with your LDAP directory will require additional configuration parameters
 specified via environment variables. These variables collectively describe how Guacamole
 will connect to LDAP:
 
-1. `LDAP_HOSTNAME` - The base of the DN for all Guacamole users.
+1. `LDAP_HOSTNAME` - The hostname or IP address of your LDAP server.
 2. `LDAP_USER_BASE_DN` - The base of the DN for all Guacamole users.
 3. `LDAP_PORT` - The port your LDAP server listens on.  (Optional)
```
Comment (mike-jumper) [inline guacamole-docker/README.md]: Hm... What do you think about listing the optional variables in their own list? I think this is how things are organized in the manual.

File: guacamole-docker/README.md (hunk 6 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: Erg

File: guacamole-docker/README.md (hunk 7 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: Arg

Comment (mike-jumper) [inline guacamole-docker/README.md]: What do you think about letting the "Using both LDAP and a database" thing be a level 3 heading beneath the general LDAP section? Like what we did with the database initialization parts?

Seems related enough that it maybe shouldn't be its own thing.

File: guacamole-docker/README.md (hunk 8 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: I could do that ...  I was trying to keep things a bit more compact by grouping them together.

Comment (mike-jumper) [inline guacamole-docker/README.md]: If you think that's better, then cool. I'm more of the mind that separate would be easier to read, but I don't think this way is bad in itself.

File: guacamole-docker/README.md (hunk 7 above)
Comment (jmuehlner) [inline guacamole-docker/README.md]: Seems reasonable.

**Comment by mike-jumper** [comment]
Here ... before we go to crazy with what's better and what's missing, perhaps it might be helpful to compare against the current manual contents. Let me generate that in a readable form ...

**Comment by jmuehlner** [comment]
@mike-jumper Ok

**Comment by mike-jumper** [comment]
OK - I've copied the generated HTML here. It should render nicely on GitHub, though the email that will be sent once I post this will be virtually useless (sorry).

----

<div class="section"><div class="titlepage"><div><div><h3 class="title"><a id="guacamole-docker-ldap"></a>Deploying Guacamole with LDAP authentication</h3></div></div></div><p>Before deploying Guacamole with the intent of using LDAP for authentication,
                please ensure that you have each of the following already prepared:</p><div class="orderedlist"><ol class="orderedlist" type="1"><li class="listitem"><p>A Docker container running the <code class="systemitem">glyptodon/guacd</code>
                        image. Guacamole needs <span class="package">guacd</span> in order to function, and
                        the Guacamole Docker image depends on a linked Docker container running
                            <span class="package">guacd</span>.</p></li><li class="listitem"><p>Network access to a working LDAP server.</p></li></ol></div><div class="section"><div class="titlepage"><div><div><h4 class="title"><a id="deploying-guacamole-docker-ldap"></a>Deploying Guacamole</h4></div></div></div><p>Linking Guacamole to your LDAP directory will require additional configuration
                    parameters specified via environment variables. These variables collectively
                    describe how Guacamole will connect to LDAP:</p><div class="informaltable"><table border="1"><colgroup><col class="c1"><col class="c2"></colgroup><thead><tr><th>Variable</th><th>Description</th></tr></thead><tbody><tr><td><code class="envar">LDAP_HOSTNAME</code></td><td>
                                    <p>The hostname or IP address of your LDAP server.</p>
                                </td></tr><tr><td><code class="envar">LDAP_USER_BASE_DN</code></td><td>
                                    <p>The base of the DN for all Guacamole users. All Guacamole
                                        users that will be authenticating against LDAP must be
                                        descendents of this base DN.</p>
                                </td></tr></tbody></table></div><p>Additional optional environment variables may be used to configure the details
                    of your LDAP directory hierarchy, encryption, or to enable more flexible
                    searching for user accounts:</p><div class="informaltable"><table border="1"><colgroup><col class="c1"><col class="c2"></colgroup><thead><tr><th>Variable</th><th>Description</th></tr></thead><tbody><tr><td><code class="envar">LDAP_PORT</code></td><td>
                                    <p>The port your LDAP server listens on. By default, this
                                        will be 389 for unencrypted LDAP or LDAP using STARTTLS, and
                                        636 for LDAP over SSL (LDAPS).</p>
                                </td></tr><tr><td><code class="envar">LDAP_ENCRYPTION_METHOD</code></td><td>
                                    <p>The encryption mechanism that Guacamole should use when
                                        communicating with your LDAP server. Legal values are "none"
                                        for unencrypted LDAP, "ssl" for LDAP over SSL/TLS (commonly
                                        known as LDAPS), or "starttls" for STARTTLS. If omitted,
                                        encryption will not be used.</p>
                                </td></tr><tr><td><code class="envar">LDAP_GROUP_BASE_DN</code></td><td>
                                    <p>The base of the DN for all groups that may be referenced
                                        within Guacamole configurations using the standard
                                            <span class="property">seeAlso</span> attribute. All groups which
                                        will be used to control access to Guacamole configurations
                                        must be descendents of this base DN. <span class="emphasis"><em>If this
                                            variable is omitted, the <span class="property">seeAlso</span>
                                            attribute will have no effect on Guacamole
                                            configurations.</em></span></p>
                                </td></tr><tr><td><code class="envar">LDAP_SEARCH_BIND_DN</code></td><td>
                                    <p>The DN (Distinguished Name) of the user to bind as when
                                        authenticating users that are attempting to log in. If
                                        specified, Guacamole will query the LDAP directory to
                                        determine the DN of each user that logs in. If omitted, each
                                        user's DN will be derived directly using the base DN
                                        specified with <code class="envar">LDAP_USER_BASE_DN</code>.</p>
                                </td></tr><tr><td><code class="envar">LDAP_SEARCH_BIND_PASSWORD</code></td><td>
                                    <p>The password to provide to the LDAP server when binding as
                                            <code class="envar">LDAP_SEARCH_BIND_DN</code> to authenticate other
                                        users. This variable is only used if
                                            <code class="envar">LDAP_SEARCH_BIND_DN</code> is specified. If
                                        omitted, but <code class="envar">LDAP_SEARCH_BIND_DN</code> is
                                        specified, Guacamole will attempt to bind with the LDAP
                                        server without a password.</p>
                                </td></tr><tr><td><code class="envar">LDAP_USERNAME_ATTRIBUTE</code></td><td>
                                    <p>The attribute or attributes which contain the username
                                        within all Guacamole user objects in the LDAP directory.
                                        Usually, and by default, this will simply be
                                            "<span class="property">uid</span>". If your LDAP directory
                                        contains users whose usernames are dictated by different
                                        attributes, multiple attributes can be specified here,
                                        separated by commas, but beware: <span class="emphasis"><em>doing so requires
                                            that a search DN be provided with
                                                <code class="envar">LDAP_SEARCH_BIND_DN</code></em></span>.</p>
                                </td></tr><tr><td><code class="envar">LDAP_CONFIG_BASE_DN</code></td><td>
                                    <p>The base of the DN for all Guacamole configurations. If
                                        omitted, the configurations of Guacamole connections will
                                        simply not be queried from the LDAP directory, and you will
                                        need to store them elsewhere, such as within a MySQL or
                                        PostgreSQL database.</p>
                                </td></tr></tbody></table></div><p>Once your <span class="package">guacd</span> container is ready, and the values of the
                    above variables are known, Guacamole can be deployed through Docker:</p><div class="informalexample"><pre class="screen"><code class="prompt">$</code> <span class="command"><strong>docker</strong></span> run --name <em class="replaceable"><code>some-guacamole</code></em> --link <em class="replaceable"><code>some-guacd</code></em>:guacd \
    -e LDAP_HOSTNAME=<em class="replaceable"><code>172.17.42.1</code></em>                            \
    -e LDAP_USER_BASE_DN=<em class="replaceable"><code>ou=people,dc=example,dc=com</code></em>        \
    -e LDAP_CONFIG_BASE_DN=<em class="replaceable"><code>ou=connections,dc=example,dc=com</code></em> \
    -d -p 8080:8080 glyptodon/guacamole</pre></div><p>As documented in <a class="xref" href="//guacamole.incubator.apache.org/doc/gug/ldap-auth.html" title="Chapter&nbsp;7.&nbsp;LDAP authentication">Chapter&nbsp;7, <em>LDAP authentication</em></a>, Guacamole does support combining LDAP with a MySQL or
                    PostgreSQL database, and this can be configured with the Guacamole Docker image,
                    as well. By providing the required environment variables for both systems,
                    Guacamole will automatically be configured to use both when the Docker image
                    starts:</p><div class="informalexample"><pre class="screen"><code class="prompt">$</code> <span class="command"><strong>docker</strong></span> run --name <em class="replaceable"><code>some-guacamole</code></em> --link <em class="replaceable"><code>some-guacd</code></em>:guacd \
    -e LDAP_HOSTNAME=<em class="replaceable"><code>172.17.42.1</code></em>                            \
    -e LDAP_USER_BASE_DN=<em class="replaceable"><code>ou=people,dc=example,dc=com</code></em>        \
    -e LDAP_CONFIG_BASE_DN=<em class="replaceable"><code>ou=connections,dc=example,dc=com</code></em> \
    <span class="emphasis"><em>-e MYSQL_HOSTNAME=<em class="replaceable"><code>172.17.42.1</code></em>   \
    -e MYSQL_DATABASE=<em class="replaceable"><code>guacamole_db</code></em>  \
    -e MYSQL_USER=<em class="replaceable"><code>guacamole_user</code></em>    \
    -e MYSQL_PASSWORD=<em class="replaceable"><code>some_password</code></em> \</em></span>
    -d -p 8080:8080 glyptodon/guacamole</pre></div><p>If any required environment variables are omitted, you will receive an error
                    message in the logs, and the image will stop. You will then need to recreate the
                    container with the proper variables specified.</p></div><div class="section"><div class="titlepage"><div><div><h4 class="title"><a id="verifying-guacamole-auth-ldap"></a>Verifying the Guacamole install</h4></div></div></div><p>Now that the Guacamole image is running, Guacamole should be accessible at
                            <code class="uri">http://<em class="replaceable"><code>HOSTNAME</code></em>:8080/guacamole/</code>,
                    where <em class="replaceable"><code>HOSTNAME</code></em> is the hostname or address of the
                    machine hosting Docker.</p><p>If you cannot access Guacamole, check the logs using Docker to determine if
                    something is wrong. Configuration parameters may have been given incorrectly, or
                    the database may be improperly initialized:</p><div class="informalexample"><pre class="screen"><code class="prompt">$</code> <span class="command"><strong>docker</strong></span> logs <em class="replaceable"><code>some-guacamole</code></em></pre></div><p>If Guacamole has been successfully installed, you will see the Guacamole login
                    screen, and should be able to login using an LDAP account beneath
                        <code class="envar">LDAP_USER_BASE_DN</code>. You will be able to access any connections
                    defined within <code class="envar">LDAP_CONFIG_BASE_DN</code> for which your user has
                    access. If a MySQL or PostgreSQL database has also been configured, and your
                    user has an account within that database, you will also be able to access your
                    connections from that database.</p></div></div>

**Comment by necouchman** [comment]
Ping.
@jmuehlner: What else needs to be done to finish this up and merge it in?

**Comment by jmuehlner** [comment]
@necouchman Well, we'd need to fix the conflicts, obviously, and take another pass at refining the documentation. When I attempted to add this I realized that I didn't have quite enough experience with LDAP to really write sufficiently useful documentation, which is why this ended up getting shelved. I probably should have closed the PR or at least made a comment to that effect...

**Comment by jmuehlner** [comment]
I'll close this one for now, since I don't have any immediate way to complete it, and it's not doing any good sitting here.
